src/node_process_methods.cc

```diff
@@ -471,6 +473,128 @@ static void ReallyExit(const FunctionCallbackInfo<Value>& args) {
+
+  CHECK_LT(length, INT_MAX);
+
+  target = new char*[length + 1];
+  target[length] = nullptr;
+
+  if (length > 0) {
+    for (int i = 0; i < length; i++) {
+      node::Utf8Value str(env->isolate(),
+                          js_array->Get(context, i).ToLocalChecked());
+      target[i] = strdup(*str);
+      CHECK_NOT_NULL(target[i]);
+    }
+  }
+
+  if (target_length != nullptr) {
+    *target_length = length;
+  }
+
+  return target;
+}
+
+#if !defined(_WIN32)
+inline int persist_standard_stream(int fd) {
+  int flags = fcntl(fd, F_GETFD, 0);
+
+  if (flags < 0) {
+    return flags;
+  }
+
+  flags &= ~FD_CLOEXEC;
+  return fcntl(fd, F_SETFD, flags);
+}
+
+static void Replace(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Local<Context> context = env->context();
+
+  CHECK_GT(args.Length(), 1);
+  CHECK(args[0]->IsString());
+
+  // Prepare arguments and environment
+  char** argv = nullptr;
+  char** envp = nullptr;
+  int envp_length = 0;
+
+  // Prepare arguments - Note that the pathname is always the first argument
+  if (args.Length() > 1) {
+    CHECK(args[1]->IsArray());
+
+    Local<Array> argv_array = args[1].As<Array>();
+
+    Local<Array> full_argv_array =
+        Array::New(env->isolate(), argv_array->Length() + 1);
+    full_argv_array->Set(context, 0, args[0].As<String>()).Check();
+    for (unsigned int i = 0; i < argv_array->Length(); i++) {
+      full_argv_array
+          ->Set(context, i + 1, argv_array->Get(context, i).ToLocalChecked())
+          .Check();
+    }
+
+    argv = copy_js_strings_array(env, full_argv_array, nullptr);
+  } else {
+    node::Utf8Value pathname_string(env->isolate(), args[0].As<String>());
+
+    argv = new char*[2];
+    argv[0] = strdup(*pathname_string);
+
+    // String duplication failed. Free up memory, then throw.
+    if (argv[0] == nullptr) {
+      int error_code = errno;
+      delete[] argv;
+      THROW_ERR_PROCESS_REPLACE_FAILED(env, error_code);
+    }
+
+    argv[1] = nullptr;
+  }
+
+  if (args.Length() > 2) {
+    CHECK(args[2]->IsArray());
+    envp = copy_js_strings_array(env, args[2].As<Array>(), &envp_length);
+  }
+
+  // Set stdin, stdout and stderr to be non-close-on-exec
+  // so that the new process will inherit it.
+  for (int i = 0; i < 3; i++) {
+    // Operation failed. Free up memory, then throw.
+    if (persist_standard_stream(i) < 0) {
+      int error_code = errno;
+      delete[] argv;
+
+      if (envp_length > 0) {
+        delete[] envp;
+      }
+
+      THROW_ERR_PROCESS_REPLACE_FAILED(env, error_code);
+    }
+  }
+
+  // Perform the execve operation.
+  // If it returns, it means that the execve operation failed.
+  RunAtExit(env);
+  execve(argv[0], argv, envp);
+  int error_code = errno;
+
+  // Free up memory, then throw.
+  delete[] argv;
+
+  if (envp_length > 0) {
+    delete[] envp;
+  }
+
+  THROW_ERR_PROCESS_REPLACE_FAILED(env, error_code);
+}
+#endif
```

Comment:
This is quite hard-to-follow C++ with lots of unnecessary explicit memory management that Node.js has been doing a lot of effort to move away from. I'd recommend looking a bit a how other parts of the Node.js code base handle strings and conversion between C++ and JS values.

Reply:
+1

Reply:
I'm absolutely willing to. Since I'm not familiar with the C++ codebase, do you have any suggestion of places I can look to?

Reply:
@ShogunPanda Well, mostly anywhere else works. As a general rule, you'll want to get rid of `new`, `new[]`, `char*`, `memcpy()` and `strdup()` as much as possible, and replace them with `std::vector`, `std::string`/`Utf8Value` as much as possible. 

(You won't be entirely able to avoid something like `std::vector<char*>` because `execve` expects `char**` arguments, but the `std::vector<char*>`'s entries could point to the entries of a `std::vector<std::string>` or `std::vector<Utf8Value>` rather than having to manage memory manually).

Reply:
> So I have to use two vectors for `argv` and two for `envp`. Is this the right approach or am I still missing something?

It can be two vectors or a single one, that doesn't really matter much 🙂

Reply:
Lol. Makes sense.
But how do you go from `std::vector<Utf8Value>` to `char**` without a second array?
I'll probably won't change the code anyway, I'm just trying to learn.